src/ccl_cls.c

```diff
@@ -137,22 +137,89 @@ static double cl_integrand(double lk, void *params) {
   return k*pk*d1*d2;
 }
 
-void ccl_angular_cls_limber(ccl_cosmology *cosmo,
-                           ccl_cl_tracer_collection_t *trc1,
-                           ccl_cl_tracer_collection_t *trc2,
-                           ccl_f2d_t *psp,
-                           int nl_out, double *l_out, double *cl_out,
-                           int *status) {
-  int local_status;
-  int clastatus, lind;
-  integ_cl_par ipar;
-  gsl_function F;
-  double lkmin, lkmax, l;
-  double result, eresult;
+static void integ_cls_limber_spline(ccl_cosmology *cosmo,
+				    integ_cl_par *ipar,
+				    double lkmin, double lkmax,
+				    double *result, int *status) {
+  int ik;
+  int nk = (int)(fmax((lkmax - lkmin) / cosmo->spline_params.DLOGK_INTEGRATION + 0.5,
+		      1))+1;
+  double *fk_arr = NULL;
+  double *lk_arr = NULL;
+  lk_arr = ccl_linear_spacing(lkmin, lkmax, nk);
```

Comment:
Is linear the best here? I don't actually know. Asking for a friend.

Reply:
The original qag integrator is solving the integral in log(k) too. If you look at the integrand, it is a lot smoother in logk than in k (that's honestly all the thinking that has gone into it).

Reply:
ohhhhh darn. I read `ccl_linear_spacing` and assumed it was linear, but you are feeding in logs! My bad! Sorry!